app/scripts/lib/rpc-method-middleware/handlers/wallet-createSession/handler.ts

```diff
@@ -117,55 +143,150 @@ async function walletCreateSessionHandler(
+      ) ||
+      Object.keys(optionalScopesWithSupportedMethodsAndNotifications).some(
+        (key) => {
+          let namespace = '';
+          try {
+            namespace = parseCaipChainId(key as CaipChainId).namespace;
+          } catch (err) {
+            return false;
+          }
+          return namespace === KnownCaipNamespace.Solana;
+        },
+      );
+
+    let promptToCreateSolanaAccount = false;
+    if (isSolanaRequested) {
+      const supportedSolanaAccounts = await hooks.getNonEvmAccountAddresses(
+        MultichainNetwork.Solana,
+      );
+      promptToCreateSolanaAccount = supportedSolanaAccounts.length === 0;
+    }
+
     const { supportedScopes: supportedRequiredScopes } = bucketScopes(
       requiredScopesWithSupportedMethodsAndNotifications,
       {
-        isChainIdSupported: networkClientExistsForChainId,
-        isChainIdSupportable: () => false, // intended for future usage with eip3085 scopedProperties
+        isEvmChainIdSupported: networkClientExistsForChainId,
+        isEvmChainIdSupportable: () => false, // intended for future usage with eip3085 scopedProperties
+        getNonEvmSupportedMethods: hooks.getNonEvmSupportedMethods,
+        isNonEvmScopeSupported: hooks.isNonEvmScopeSupported,
       },
     );
 
     const { supportedScopes: supportedOptionalScopes } = bucketScopes(
       optionalScopesWithSupportedMethodsAndNotifications,
       {
-        isChainIdSupported: networkClientExistsForChainId,
-        isChainIdSupportable: () => false, // intended for future usage with eip3085 scopedProperties
+        isEvmChainIdSupported: networkClientExistsForChainId,
+        isEvmChainIdSupportable: () => false, // intended for future usage with eip3085 scopedProperties
+        getNonEvmSupportedMethods: hooks.getNonEvmSupportedMethods,
+        isNonEvmScopeSupported: hooks.isNonEvmScopeSupported,
       },
     );
 
+    const allScopesRequested = Object.keys(supportedRequiredScopes).concat(
+      Object.keys(supportedOptionalScopes),
+    );
+
     // Fetch EVM accounts from native wallet keyring
-    // These addresses are lowercased already
     const existingEvmAddresses = hooks
       .listAccounts()
       .map((account) => account.address);
-    const supportedEthAccounts = getEthAccounts({
-      requiredScopes: supportedRequiredScopes,
-      optionalScopes: supportedOptionalScopes,
-    })
-      .map((address) => address.toLowerCase() as Hex)
-      .filter((address) => existingEvmAddresses.includes(address));
+
+    // TODO dry and or move to @metamask/chain-agnostic-permission
+    const requestedRequiredAccounts = Object.values(
+      supportedRequiredScopes,
+    ).flatMap((scope) => scope.accounts);
+    const requestedOptionalAccounts = Object.values(
+      supportedOptionalScopes,
+    ).flatMap((scope) => scope.accounts);
+
+    const allRequestedAccountAddresses = uniq([
+      ...requestedRequiredAccounts,
+      ...requestedOptionalAccounts,
+    ]);
```

Comment:
replace this with getPermittedAccounts helper

Reply:
Working on this